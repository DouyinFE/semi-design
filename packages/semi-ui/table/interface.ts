/* eslint-disable max-len */
import React, { ReactNode, MutableRefObject } from 'react';

import { BaseProps } from '../_base/baseComponent';
import { PaginationProps } from '../pagination';
import { CheckboxProps } from '../checkbox';
import { DropdownProps } from '../dropdown';
import { Locale } from '../locale/interface';
import { ArrayElement } from '../_base/base';
import { strings } from '@douyinfe/semi-foundation/table/constants';
import {
    BaseRowKeyType,
    BaseSortOrder,
    BaseGroupBy,
    BaseGroupByFn,
    BaseFixed,
    BaseAlign,
    BaseChangeInfoSorter,
    BaseSorter,
    BaseFilter,
    BaseChangeInfoFilter,
    BaseIncludeGroupRecord
} from '@douyinfe/semi-foundation/table/foundation';
import { ScrollDirection, CSSDirection } from 'react-window';

export interface TableProps<RecordType extends Record<string, any> = any> extends BaseProps {
    bordered?: boolean;
    children?: ReactNode;
    childrenRecordName?: string;
    className?: string;
    clickGroupedRowToExpand?: boolean;
    columns?: ColumnProps<RecordType>[];
    components?: TableComponents;
    dataSource?: RecordType[];
    defaultExpandAllGroupRows?: boolean;
    defaultExpandAllRows?: boolean;
    defaultExpandedRowKeys?: (string | number)[];
    empty?: ReactNode;
    expandAllGroupRows?: boolean;
    expandAllRows?: boolean;
    expandCellFixed?: Fixed;
    expandIcon?: ExpandIcon;
    expandedRowKeys?: (string | number)[];
    expandedRowRender?: ExpandedRowRender<RecordType>;
    expandRowByClick?: boolean;
    footer?: Footer<RecordType>;
    getVirtualizedListRef?: GetVirtualizedListRef;
    groupBy?: GroupBy<RecordType>;
    hideExpandedColumn?: boolean;
    id?: string;
    indentSize?: number;
    loading?: boolean;
    pagination?: TablePagination;
    prefixCls?: string;
    renderGroupSection?: RenderGroupSection;
    renderPagination?: RenderPagination;
    resizable?: Resizable<RecordType>;
    rowExpandable?: RowExpandable<RecordType>;
    rowKey?: RowKey<RecordType>;
    rowSelection?: RowSelection<RecordType>;
    scroll?: Scroll;
    showHeader?: boolean;
    size?: Size;
    style?: React.CSSProperties;
    title?: Title<RecordType>;
    virtualized?: Virtualized;
    onChange?: OnChange<RecordType>;
    onExpand?: OnExpand<RecordType>;
    onExpandedRowsChange?: OnExpandedRowsChange<RecordType>;
    onGroupedRow?: OnGroupedRow<RecordType>;
    onHeaderRow?: OnHeaderRow<RecordType>;
    onRow?: OnRow<RecordType>;
}

export interface ColumnProps<RecordType extends Record<string, any> = any> {
    [x: string]: any;
    align?: Align;
    children?: Array<ColumnProps<RecordType>>;
    className?: string;
    colSpan?: number;
    dataIndex?: string;
    defaultSortOrder?: SortOrder;
    filterChildrenRecord?: boolean;
    filterDropdown?: React.ReactNode;
    filterDropdownProps?: DropdownProps;
    filterDropdownVisible?: boolean;
    filterIcon?: FilterIcon;
    filterMultiple?: boolean;
    filteredValue?: any[];
    filters?: Filter[];
    fixed?: Fixed;
    key?: string | number;
    render?: ColumnRender<RecordType>;
    renderFilterDropdownItem?: RenderFilterDropdownItem;
    sortChildrenRecord?: boolean;
    sortOrder?: SortOrder;
    sorter?: Sorter<RecordType>;
    title?: ColumnTitle;
    useFullRender?: boolean;
    width?: string | number;
    onCell?: OnCell<RecordType>;
    onFilter?: OnFilter<RecordType>;
    onFilterDropdownVisibleChange?: OnFilterDropdownVisibleChange;
    onHeaderCell?: OnHeaderCell<RecordType>;
}

export type Align = BaseAlign;
export type SortOrder = BaseSortOrder;
export type FilterIcon = boolean | React.ReactNode | FilterIconRenderFunction;
export interface Filter extends BaseFilter {
    value?: any;
    text?: React.ReactNode;
    children?: Filter[];
}
export type Fixed = BaseFixed;
export type OnCell<RecordType> = (record?: RecordType, rowIndex?: number) => OnCellReturnObject;
export type OnFilter<RecordType> = (filteredValue?: any, record?: RecordType) => boolean;
export type OnFilterDropdownVisibleChange = (visible?: boolean) => void;
export type OnHeaderCell<RecordType> = (record?: RecordType, columnIndex?: number) => OnHeaderCellReturnObject;
export type ColumnRender<RecordType> = (text: any, record: RecordType, index: number, options?: RenderOptions) => ColumnRenderReturnType;
export type RenderFilterDropdownItem = (itemInfo?: FilterDropdownItem) => ReactNode;
export type Sorter<RecordType> = BaseSorter<RecordType>;
export type ColumnTitle = React.ReactNode | ((ColumnTitleProps?: ColumnTitleProps) => React.ReactNode);
export type FilterIconRenderFunction = (filtered: boolean) => React.ReactNode;
export type ColumnTitleProps = {
    sorter?: React.ReactNode;
    filter?: React.ReactNode;
    selection?: React.ReactNode;
};
export type ColumnRenderReturnType = React.ReactNode | RenderReturnObject;
export interface RenderReturnObject {
    [x: string]: any;
    children: React.ReactNode;
    props: {
        [x: string]: any;
        colSpan?: number;
        rowSpan?: number;
    };
}
export interface FilterDropdownItem {
    [x: string]: any;
    value?: any;
    text?: React.ReactNode;
    onChange?: React.MouseEventHandler<HTMLLIElement>;
    level?: number;
    filterMultiple?: boolean;
    checked?: boolean;
}
export interface RenderOptions {
    expandIcon?: React.ReactNode;
}
export interface OnCellReturnObject extends React.TdHTMLAttributes<HTMLElement> {
    [x: string]: any;
    style?: React.CSSProperties;
    className?: string;
    onClick?: (e: React.MouseEvent) => void;
}
export interface OnHeaderCellReturnObject extends React.ThHTMLAttributes<HTMLElement> {
    [x: string]: any;
    style?: React.CSSProperties;
    className?: string;
    onClick?: (e: React.MouseEvent) => void;
}

interface OnRowReturnOmit {
    ref?: React.RefObject<any>;
}

export interface OnRowReturnObject extends Omit<React.DetailedHTMLProps<React.HTMLAttributes<HTMLTableRowElement>, HTMLTableRowElement>, keyof OnRowReturnOmit> {
    [x: string]: any;
    className?: string;
    style?: React.CSSProperties;
    onClick?: (e: React.MouseEvent) => void;
}
export interface OnGroupedRowReturnObject extends Omit<React.HTMLAttributes<HTMLTableRowElement>, 'className'> {
    [x: string]: any;
    style?: React.CSSProperties;
    onClick?: (e: React.MouseEvent) => void;
}
export type OnHeaderRowReturnObject = Omit<React.HTMLAttributes<HTMLTableRowElement>, 'ref' | 'style'>;

export interface Scroll {
    x?: number | string;
    y?: number | string;
    scrollToFirstRowOnChange?: boolean;
}

export interface Data {
    [x: string]: any;
    key?: string | number;
}

export interface TableComponents {
    table?: ReactNode;
    header?: {
        outer?: ReactNode;
        wrapper?: ReactNode;
        row?: ReactNode;
        cell?: ReactNode;
    };
    body?: {
        outer?: ReactNode;
        wrapper?: ReactNode;
        row?: ReactNode;
        cell?: ReactNode;
        colgroup?: {
            wrapper?: ReactNode;
            col?: ReactNode;
        };
    };
    footer?: {
        wrapper?: ReactNode;
        row?: ReactNode;
        cell?: ReactNode;
        outer?: ReactNode;
    };
}

export interface RowSelectionProps<RecordType> {
    className?: string;
    disabled?: boolean;
    fixed?: Fixed;
    getCheckboxProps?: GetCheckboxProps<RecordType>;
    hidden?: boolean;
    selectedRowKeys?: (string | number)[];
    title?: ReactNode;
    width?: string | number;
    onChange?: RowSelectionOnChange<RecordType>;
    onSelect?: RowSelectionOnSelect<RecordType>;
    onSelectAll?: RowSelectionOnSelectAll<RecordType>;
}

export type GetCheckboxProps<RecordType> = (record: RecordType) => CheckboxProps;
export type RowSelectionOnChange<RecordType> = (selectedRowKeys?: (string | number)[], selectedRows?: RecordType[]) => void;
export type RowSelectionOnSelect<RecordType> = (
    record?: RecordType,
    selected?: boolean,
    selectedRows?: RecordType[],
    nativeEvent?: React.MouseEvent
) => void;
export type RowSelectionOnSelectAll<RecordType> = (selected?: boolean, selectedRows?: RecordType[], changedRows?: RecordType[]) => void;
export type ExpandIcon = ((expanded?: boolean) => React.ReactNode) | React.ReactNode | boolean;
export type ExpandedRowRender<RecordType> = (record?: RecordType, index?: number, expanded?: boolean) => React.ReactNode;
export type Footer<RecordType> = ReactNode | ((pageData?: RecordType[]) => React.ReactNode);
export type FormatPageText = ((pageInfo?: { currentStart?: number; currentEnd?: number; total?: number }) => React.ReactNode) | boolean;
export type GetVirtualizedListRef = (ref: MutableRefObject<any>) => void;
export type GroupByFunction<RecordType> = BaseGroupByFn<RecordType>;
export type GroupBy<RecordType> = BaseGroupBy<RecordType>;
export type Size = ArrayElement<typeof strings.SIZES>;
export type Title<RecordType> = React.ReactNode | ((pageData?: RecordType[]) => React.ReactNode);
export type PaginationPosition = ArrayElement<typeof strings.PAGINATION_POSITIONS>;
export type Pagination = TablePaginationProps | boolean;
export type TablePagination = Pagination;
export interface ChangeInfoFilter<RecordType> extends BaseChangeInfoFilter<RecordType> {
    filters?: Filter[];
    onFilter?: OnFilter<RecordType>;
}
export type ChangeInfoSorter<RecordType> = BaseChangeInfoSorter<RecordType>;
export interface ChangeInfo<RecordType> {
    pagination?: TablePaginationProps;
    filters?: ChangeInfoFilter<RecordType>[];
    sorter?: ChangeInfoSorter<RecordType>;
    extra?: Record<string, any>;
}
export type OnChange<RecordType> = (changeInfo: ChangeInfo<RecordType>) => void;
export type OnRow<RecordType> = (record?: RecordType, index?: number) => OnRowReturnObject;
export type OnGroupedRow<RecordType> = (record?: RecordType, index?: number) => OnGroupedRowReturnObject;
export type OnHeaderRow<RecordType> = (columns?: ColumnProps<RecordType>[], index?: number) => OnHeaderRowReturnObject;
export type OnExpandedRowsChange<RecordType> = (expandedRows?: IncludeGroupRecord<RecordType>[]) => void;
export type OnExpand<RecordType> = (expanded?: boolean, record?: IncludeGroupRecord<RecordType>, mouseEvent?: React.MouseEvent) => void;
export type RenderGroupSection = (groupKey?: string | number, group?: (string | number)[]) => ReactNode | {
    [x: string]: any;
    children: ReactNode;
};
export type RenderPagination = (paginationProps: TablePaginationProps) => ReactNode;
export type RowExpandable<RecordType> = (record?: RecordType) => boolean;
export type RowKey<RecordType> = BaseRowKeyType | ((record?: RecordType) => string);
export type RowSelection<RecordType> = RowSelectionProps<RecordType> | boolean;

export type VirtualizedOnScrollArgs = {
    scrollDirection?: ScrollDirection;
    scrollOffset?: number;
    scrollUpdateWasRequested?: boolean;
};
export type VirtualizedMode = 'list' | 'grid';
export type VirtualizedItemSizeFn = (index?: number) => number;
export type VirtualizedItemSize = number | VirtualizedItemSizeFn;
export type VirtualizedOnScroll = (object: VirtualizedOnScrollArgs) => void;
export interface VirtualizedProps {
    [x: string]: any;
    mode?: VirtualizedMode;
    itemSize?: VirtualizedItemSize;
    onScroll?: VirtualizedOnScroll;
}
export type Virtualized = boolean | VirtualizedProps;

export interface TablePaginationProps extends BaseProps, PaginationProps {
    position?: PaginationPosition;
    formatPageText?: FormatPageText;
}

export type Resizable<RecordType> = ResizableProps<RecordType> | boolean;
export interface ResizableProps<RecordType> {
    onResize?: ResizeFn<RecordType>;
    onResizeStart?: ResizeFn<RecordType>;
    onResizeStop?: ResizeFn<RecordType>;
}
export type ResizeFn<RecordType> = (column: RecordType) => RecordType;

export interface BodyScrollEvent extends React.UIEvent {
    [x: string]: any;
    currentTarget: any;
    target: any;
}

export type BodyScrollPosition = 'both' | 'middle' | 'left' | 'right';

export type TableLocale = Locale['Table'];
export type Direction = CSSDirection;
export type IncludeGroupRecord<RecordType> = BaseIncludeGroupRecord<RecordType>;