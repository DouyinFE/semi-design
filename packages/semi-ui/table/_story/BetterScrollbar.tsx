import React from 'react';
import { Table, Tooltip, Tag, Space, Typography, Switch } from '../../index';
import RTLWrapper from '../../configProvider/_story/RTLDirection/RTLWrapper';

function App() {
    const [bordered, setBordered] = React.useState(true);
    const columns = [
        {
            title: 'Name',
            dataIndex: 'name',
            width: 150,
            fixed: true,
            filters: [
                {
                    text: 'Code 45',
                    value: '45',
                },
                {
                    text: 'King 4',
                    value: 'King 4',
                },
            ],
            onFilter: (value, record) => record.name.includes(value),
        },
        {
            title: 'Age',
            dataIndex: 'age',
            fixed: true,
            width: 150,
            sorter: (a, b) => (a.age - b.age > 0 ? 1 : -1),
        },
        {
            title: 'Address',
            width: 200,
            dataIndex: 'address',
        },
        // {
        //     title: 'Address2',
        //     width: 200,
        //     dataIndex: 'address',
        // },
        // {
        //     title: 'Address3',
        //     width: 200,
        //     dataIndex: 'address',
        // },
        // {
        //     title: 'Address4',
        //     width: 200,
        //     dataIndex: 'address',
        // },
        // {
        //     title: 'Address5',
        //     width: 200,
        //     dataIndex: 'address',
        // },
        {
            fixed: 'right' as const,
            width: 250,
            render: (text, record) => (
                <Tooltip content={record.description}>
                    <Tag color="green">Show Info</Tag>
                </Tooltip>
            ),
        },
    ];
    
    const data = [];
    
    for (let i = 0; i < 10; i++) {
        const age = (i * 1000) % 149 ;
        const name = `Edward King ${i}`;
        data.push({
            key: `${ i}`,
            name,
            age,
            address: `London, Park Lane no. ${i}`,
            description: `My name is ${name}, I am ${age} years old, living in New York No. ${i + 1} Lake Park.`,
        });
    }
    
    const allColumnsWidth = columns.reduce((count, column) => count + column.width, 0);

    const tableProps = {
        pagination: false,
        columns,
        bordered,
        dataSource: data,
    };

    return (
        <RTLWrapper>
            <div style={{ textAlign: 'left', marginBottom: 12 }}>
                <Space>
                    <Typography.Title heading={6}>边框</Typography.Title>
                    <Switch onChange={() => setBordered(!bordered)} checked={bordered} />
                </Space>
            </div>
            <Table
                {...tableProps}
                scroll={{ y: 400 }}
            />
            <br />
            <Table
                {...tableProps}
                title="虚拟化表格"
                scroll={{ y: 400, x: allColumnsWidth }}
                style={{ width: allColumnsWidth }}
                virtualized
            />
        </RTLWrapper>
    );
}

export default App;