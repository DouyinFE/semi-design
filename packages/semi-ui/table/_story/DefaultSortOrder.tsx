import React, { useState } from 'react';
import { Table, Button } from '../../index';
import { ColumnProps } from '../interface';

interface Record {
    title?: string;
    age?: number;
    address?: string;
    name?: string
}

const columns: ColumnProps<Record>[] = [
    {
        title: 'Name',
        dataIndex: 'name',
        sorter: (a: Record, b: Record): number => a.name.length - b.name.length,
        defaultSortOrder: 'descend',
    },
    {
        title: 'Age',
        dataIndex: 'age',
        sorter: (a: Record, b: Record): number => (a.age - b.age > 0 ? 1 : -1),
        defaultSortOrder: 'ascend',
    },
    {
        title: 'Address',
        dataIndex: 'address',
        sorter: (a: Record, b: Record): number => a.address.length - b.address.length,
        defaultSortOrder: false,
    },
];

const data1 = [
    {
        key: '1',
        name: 'John Brown',
        age: 32,
        address: 'New York No. 1 Lake Park',
    },
    {
        key: '2',
        name: 'Jim Green',
        age: 42,
        address: 'London No. 1 Lake Park',
    },
    {
        key: '3',
        name: 'Joe Black',
        age: 32,
        address: 'Sidney No. 1 Lake Park',
    },
    {
        key: '4',
        name: 'Jim Red',
        age: 32,
        address: 'London No. 2 Lake Park',
    },
];

const data2 = [
    {
        key: '1',
        name: 'John Brown',
        age: 1,
        address: 'New York No. 1 Lake Park',
    },
    {
        key: '2',
        name: 'Jim Green',
        age: 2,
        address: 'London No. 1 Lake Park',
    },
    {
        key: '3',
        name: 'Joe Black',
        age: 4,
        address: 'Sidney No. 1 Lake Park',
    },
    {
        key: '4',
        name: 'Jim Red',
        age: 3,
        address: 'London No. 2 Lake Park',
    },
];

interface OnChangeArgs {
    filters?: unknown[];
    sorter?: unknown;
    extra?: unknown
}

function App(): React.ReactElement {
    const [data, setData] = useState(data1);
    const [flag, setFlag] = useState(false);
    const handleChange = ({ sorter }: OnChangeArgs): void => {
        console.log('table sorter changed to', sorter);
    };

    const handleClick = (): void => {
        const newData = flag ? data1 : data2;
        setData(newData);
        setFlag(!flag);
    };

    const currentData = flag ? 'data2' : ' data1';

    return (
        <div>
            <div>
                <label>defaultSortOrder=ascend</label>
                <Table columns={columns} dataSource={data} onChange={handleChange} />
            </div>
            <Button onClick={handleClick}>update data</Button>
            <div>{currentData}</div>
        </div>
    );
}

export default App;