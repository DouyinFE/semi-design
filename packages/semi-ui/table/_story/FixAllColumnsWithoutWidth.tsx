import React from 'react';
import { Table, Tooltip, Tag } from '../../index';

function App() {
    const columns = [
        {
            title: 'Name',
            dataIndex: 'name',
        },
        {
            title: 'Age',
            dataIndex: 'age',
        },
        {
            title: 'Address',
            dataIndex: 'address',
        },
        {
            title: 'Description',
            dataIndex: 'description',
        },
        {
            dataIndex: 'x',
            render: (text, record) => (
                <Tooltip content={record.description}>
                    <Tag color="green">Show Info</Tag>
                </Tooltip>
            ),
        },
    ];

    const data = [];

    for (let i = 0; i < 8; i++) {
        const age = (i * 1000) % 149 ;
        const name = `Edward King ${i}`;
        data.push({
            key: `${i}`,
            name,
            age,
            address: `London, Park Lane no. ${i}`,
            address2: `London, Park Lane no. ${i}`,
            description: `My name is ${name}`,
        });
    }

    const scroll = { y: 300 };

    return <Table bordered columns={columns} dataSource={data} scroll={scroll} pagination={false} />;
}

export default App;