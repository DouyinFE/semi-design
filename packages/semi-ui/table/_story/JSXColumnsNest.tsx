import React, { useMemo } from 'react';
import Table from '../index';
import { DataType } from '../Table';

const JSXColumnsNest: React.FunctionComponent = () => {
    const data = useMemo(() => {
        const _data = [];
        for (let i = 0; i < 100; i++) {
            const age = (i * 1000) % 149 ;
            const name = `Edward King ${i}`;
            _data.push({
                key: `${ i}`,
                company: {
                    name: 'ByteDance',
                    address: 'No. 48, Zhichun Road',
                },
                name,
                age,
                address: `No ${i + 1}, Zhongguancun Street`,
                description: `My name is ${name}, I am ${age} years old, living in No ${i + 1}, Zhongguancun Street`,
            });
        }
        return _data;
    }, []);

    return (
        <Table
            rowSelection={{ fixed: true }}
            expandedRowRender={(record: DataType): React.ReactNode => <article>{record.description}</article>}
            dataSource={data}
            scroll={{ x: '120%', y: 400 }}
            onChange={(...args: Array<unknown>): void => console.log(args)}
        >
            <Table.Column title={'Base Information'} fixed>
                <Table.Column
                    title={'Name'}
                    dataIndex={'name'}
                    fixed
                    width={200}
                    filters={[
                        {
                            text: 'Code 45',
                            value: '45',
                        },
                        {
                            text: 'King 4',
                            value: 'King 4',
                        },
                    ]}
                    onFilter={(value, record): boolean => record.name.includes(value)}
                />
                {
                    // 这样做是为了查看是否符合类型定义
                    (Math.random() > 0.5) ? (
                        <Table.Column
                            title={'Age'}
                            dataIndex={'age'}
                            fixed
                            width={100}
                            sorter={(a: DataType, b: DataType): number => (a.age - b.age > 0 ? 1 : -1)}
                        />
                    ) : null
                }
            </Table.Column>
            <Table.Column title={'Company Information'}>
                <Table.Column title={'Company Name'} dataIndex={'company.name'} />
                <Table.Column title={'Company Address'} dataIndex={'company.address'} />
            </Table.Column>
            <Table.Column title={'Address'} width={250} fixed={'right'} dataIndex={'address'} />
        </Table>
    );
};

export default JSXColumnsNest;