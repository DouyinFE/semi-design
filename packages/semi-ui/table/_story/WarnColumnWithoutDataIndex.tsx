import React from 'react';
import { Table, Tooltip, Tag  } from '../../index';

function App() {
    const columns = [
        {
            title: 'Name',
            width: 150,
            fixed: true,
            filters: [
                {
                    text: 'Code 45',
                    value: '45',
                },
                {
                    text: 'King 4',
                    value: 'King 4',
                },
            ],
            onFilter: (value, record) => record.name.includes(value),
        },
        {
            title: 'Age',
            fixed: true,
            width: 150,
            sorter: (a, b) => (a.age - b.age > 0 ? 1 : -1),
        },
        {
            title: 'Address',
            width: 200,
            dataIndex: 'address',
        },
        {
            fixed: 'right' as const,
            width: 250,
            render: (text, record) => (
                <Tooltip content={record.description}>
                    <Tag color="green">Show Info</Tag>
                </Tooltip>
            ),
        },
    ];
    
    const data = [];
    
    for (let i = 0; i < 10; i++) {
        const age = 40 + (Math.random() > 0.5 ? 1 : -1) * (i % 9);
        const name = `Edward King ${i}`;
        data.push({
            key: `${ i}`,
            name,
            age,
            address: `London, Park Lane no. ${i}`,
            description: `My name is ${name}, I am ${age} years old, living in New York No. ${i + 1} Lake Park.`,
        });
    }
    
    const tableProps = {
        pagination: false,
        columns,
        dataSource: data,
    };

    return (
        <Table
            {...tableProps}
        />
    );
}

export default App;