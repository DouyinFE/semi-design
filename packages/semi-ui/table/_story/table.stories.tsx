import React, { ReactNode, useMemo } from 'react';
import { storiesOf } from '@storybook/react';
import { IllustrationConstruction } from '@douyinfe/semi-illustrations';

import { Fixed, ColumnProps, OnRow, OnHeaderRow, OnGroupedRow, RowKey } from '../interface';
import JSXColumnsNest from './JSXColumnsNest';
import DefaultSortOrder from './DefaultSortOrder';
import BetterScrollbar from './BetterScrollbar';
import Empty from '../../empty';
import Table from '../index';
import FixAllColumnsWithoutWidth from './FixAllColumnsWithoutWidth';
import WarnColumnWithoutDataIndex from './WarnColumnWithoutDataIndex';

const stories = storiesOf('Table', module);

interface MyData {
    title?: string;
    dataIndex?: string;
    width?: number;
    render?: Function;
    key?: string;
    name?: string;
    age?: number;
    address?: string;
}

const columns: ColumnProps<MyData>[] = [
    {
        title: 'Name',
        dataIndex: 'name',
        width: 200,
        render: (text: ReactNode) => <a>{text}</a>,
        onCell: (record: MyData, index) => {
            console.log(record.title, index);
            return {};
        },
    },
    {
        title: 'Age',
        width: 200,
        dataIndex: 'age',
    },
    {
        title: 'Address',
        dataIndex: 'address',
    },
];
const data: MyData[] = [
    {
        key: '1',
        name: 'John Brown',
        age: 32,
        address: 'New York No. 1 Lake Park, New York No. 1 Lake Park',
    },
    {
        key: '2',
        name: 'Jim Green',
        age: 42,
        address: 'London No. 1 Lake Park',
    },
    {
        key: '3',
        name: 'Joe Black',
        age: 32,
        address: 'Sidney No. 1 Lake Park',
    },
    {
        key: '4',
        name: 'Disabled User',
        age: 99,
        address: 'Sidney No. 1 Lake Park',
    },
];

stories.add('basic table', () => <Table bordered resizable columns={columns} dataSource={data} />);

stories.add(`jsx table`, () => (
    <Table dataSource={data} pagination={false}>
        <Table.Column
            title="Name"
            dataIndex="name"
            key="name"
            render={(text: string) => <a>{text}</a>}
            filters={[
                {
                    text: 'London',
                    value: 'London',
                },
                {
                    text: 'New York',
                    value: 'New York',
                },
            ]}
        />
        <Table.Column title="Age" dataIndex="age" key="age" />
        <Table.Column title="Address" dataIndex="address" key="address" />
    </Table>
));

stories.add(`jsx nested cloumn`, () => <JSXColumnsNest />);

stories.add(`header merge table`, () => {
    const data = useMemo(() => {
        const data = [];
        for (let i = 0; i < 100; i++) {
            let age = (i * 1000) % 149 ;
            let name = `Edward King ${i}`;
            data.push({
                key: '' + i,
                company: {
                    name: 'ByteDance',
                    address: 'No. 48, Zhichun Road',
                },
                name,
                age,
                address: `No ${i + 1}, Zhongguancun Street`,
                description: `My name is ${name}, I am ${age} years old, living in No ${i + 1}, Zhongguancun Street`,
            });
        }
        return data;
    }, []);

    const columns = useMemo(
        () => [
            {
                title: 'Base Information',
                fixed: 'left' as Fixed,
                children: [
                    {
                        title: 'Name',
                        dataIndex: 'name',
                        fixed: 'left' as Fixed,
                        width: 200,
                        filters: [
                            {
                                text: 'Code 45',
                                value: '45',
                            },
                            {
                                text: 'King 4',
                                value: 'King 4',
                            },
                        ],
                        onFilter: (value: any, record: any) => record.name.includes(value),
                    },
                    {
                        title: 'Age',
                        dataIndex: 'age',
                        fixed: 'left' as Fixed,
                        width: 100,
                        sorter: (a: any, b: any) => (a.age - b.age > 0 ? 1 : -1),
                    },
                ],
            },
            {
                title: 'Company Information',
                children: [
                    {
                        title: 'Company Name',
                        dataIndex: 'company.name',
                    },
                    {
                        title: 'Company Address',
                        dataIndex: 'company.address',
                    },
                ],
            },
            {
                title: 'Address',
                width: 250,
                dataIndex: 'address',
                fixed: 'right' as Fixed,
            },
        ],
        []
    );

    return (
        <Table
            rowSelection={{ fixed: true, getCheckboxProps: (record) => ({ disabled: record.name === 'semi' }) }}
            expandedRowRender={(record: any) => <article>{record.description}</article>}
            dataSource={data}
            scroll={{ x: '120%', y: 400 }}
            onChange={(...args: any) => console.log(args)}
            columns={columns}
        />
    );
});

stories.add('table generic', () => (
    <Table<MyData>
        columns={columns}
        dataSource={data}
        rowExpandable={record => {
            return record.age > 30 ? true : false;
        }}
        onHeaderRow={() => ({ onClick: () => {}})}
    />
));

stories.add('table column generic', () => <Table columns={columns} dataSource={data} />);
stories.add('defaultSortOrder', () => <DefaultSortOrder />);

stories.add('expandRowByClick', () => (
    <Table<MyData>
        expandRowByClick
        onExpand={(...args) => {
            console.log('click row', ...args);
        }}
        onExpandedRowsChange={rows => {
            console.log('rows change', rows);
        }}
        expandedRowRender={() => <div>Semi Design</div>}
        columns={columns}
        dataSource={data}
    />
));

stories.add('onCell/onHeaderCell', () => {
    const columns: ColumnProps<MyData>[] = [
        {
            title: 'Name',
            dataIndex: 'name',
            width: 200,
            render: (text: ReactNode) => <a>{text}</a>,
            onCell: (record: MyData, index: number) => {
                console.log(`current record name is ${record.title}; row key is ${index}`);
                return index % 2 === 0 ? ({
                    style: { background: 'red' },
                    onClick: () => console.log(`click ${index}`),
                    onMouseEnter: () => console.log(`mouse enter ${index}`),
                    align: 'right',
                }) : ({});
            },
        },
        {
            title: 'Age',
            width: 200,
            dataIndex: 'age',
            onHeaderCell: (record: MyData, index) => {
                console.log(`current record name is ${record.title}; row key is ${index}`);
                return ({
                    style: { background: 'blue' },
                    onClick: () => console.log(`click ${index}`),
                    onMouseLeave: () => console.log(`mouse leave ${index}`),
                    className: `test-header-cell-${index}`,
                });
            },
        },
        {
            title: 'Address',
            dataIndex: 'address',
        },
    ];
    return (
        <Table columns={columns} dataSource={data} />
    )
});

stories.add('onRow/onHeaderRow', () => {
    const onRow: OnRow<any> = (record, index) => {
        if (index % 2 === 0) {
            return ({
                style: {
                    background: 'var(--semi-color-fill-2)',
                },
                onClick: () => {
                    console.log(`click row ${index}`);
                }
            });
        }
    };
    const onHeaderRow: OnHeaderRow<any> = (record, index) => {
        return ({
            onClick: () => {
                console.log(`click header row ${index}`);
            },
            className: 'test-header-row',
        });
    };
    const columns: ColumnProps<MyData>[] = [
        {
            title: 'Name',
            dataIndex: 'name',
            width: 200,
            render: (text: ReactNode) => <a>{text}</a>,
        },
        {
            title: 'Age',
            width: 200,
            dataIndex: 'age',
        },
        {
            title: 'Address',
            dataIndex: 'address',
        },
    ];
    return (
        <Table columns={columns} dataSource={data} onRow={onRow} onHeaderRow={onHeaderRow} />
    )
});

stories.add('onGroupedRow', () => {
    interface DataItem {
        city: string;
        job: string;
        department: string;
    }
    const rowKey: RowKey<DataItem> = record => `${record.city && record.city.toLowerCase()}-${record.job && record.job.toLowerCase()}`;
    const data: DataItem[] = [
        { city: 'Beijing', job: 'FE', department: 'IES' },
        { city: 'Beijing', job: 'BE', department: 'IES' },
        { city: 'Shanghai', job: 'Android', department: 'IES' },
        { city: 'Tokyo', job: 'Android', department: 'IES' },
        { city: 'Shanghai', job: 'IOS', department: 'EE' },
        { city: 'LA', job: 'SE', department: 'EE' },
        { city: 'Beijing', job: 'Android', department: 'EE' },
        { city: 'Tokyo', job: 'IOS', department: 'EE' },
        { city: 'Tokyo', job: 'SE', department: 'DATA' },
        { city: 'Shanghai', job: 'BE', department: 'DATA' },
        { city: 'LA', job: 'Android', department: 'DATA' },
        { city: 'LA', job: 'IOS', department: 'DATA' },
    ];

    const columns: ColumnProps<DataItem>[] = [
        { dataIndex: 'city', title: 'City', width: 400, sorter: (a, b) => (a.city > b.city ? 1 : -1) },
        {
            dataIndex: 'job',
            title: 'Job',
            width: 200,
            filters: [
                { text: 'IOS', value: 'IOS' },
                { text: 'Android', value: 'Android' },
            ],
            onFilter: (value, record) => record.job && record.job.indexOf(value) === 0,
        },
        { dataIndex: 'department', title: 'Department' },
    ];

    return (
        <div style={{ padding: '20px 0px' }}>
            <Table<DataItem>
                dataSource={data}
                rowKey={rowKey}
                groupBy={'city'}
                columns={columns}
                renderGroupSection={groupKey => <strong>Jobs in {groupKey}:</strong>}
                onGroupedRow={(group, index) => {
                    return {
                        onClick: e => { console.log(`Grouped row clicked: `, group, index) },
                        style: { color: 'var(--semi-color-primary)' },
                    };
                }}
                clickGroupedRowToExpand
                scroll={{ y: 480 }}
            />
        </div>
    );
});

stories.add('empty', () => {
    const empty = 'this is a empty placeholder';
    const test = (
        <div>
            <Empty
                image={IllustrationConstruction}
                title={'功能建设中'}
                description="当前功能暂未开放，敬请期待。"
            />
        </div>
    );

    return (
        <Table columns={columns} dataSource={[]} empty={test} />
    );
});

stories.add('better scrollbar', () => <BetterScrollbar />);

stories.add('fix all columns without width', () => <FixAllColumnsWithoutWidth />);

stories.add('warn if column without dataIndex', () => <WarnColumnWithoutDataIndex />);