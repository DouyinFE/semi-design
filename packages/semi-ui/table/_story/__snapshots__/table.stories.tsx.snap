// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic table 1`] = `null`;

exports[`better scrollbar 1`] = `null`;

exports[`defaultSortOrder 1`] = `null`;

exports[`empty 1`] = `null`;

exports[`expandRowByClick 1`] = `null`;

exports[`fix all columns without width 1`] = `null`;

exports[`jsx nested cloumn 1`] = `null`;

exports[`jsx table 1`] = `null`;

exports[`onCell/onHeaderCell 1`] = `null`;

exports[`onGroupedRow 1`] = `null`;

exports[`onRow/onHeaderRow 1`] = `null`;

exports[`table column generic 1`] = `null`;

exports[`table generic 1`] = `null`;

exports[`warn if column without dataIndex 1`] = `null`;
